### PR TITLE
Plotting Components Plugin

### PR DESCRIPTION
**Pull request by AmrElsersy**
A plugin for handling **Components Plotting** that uses PlottingInterface
this includes

- Making the ComponentInspector plugin items are dragable to be dragged into the plotting interface

- Making a gazebo-plugin to handle charts subscription to that components and updates the plotting interface

Screenshot from 2020-08-03 18-45-46

This class diagram shows the design and the integration between it and plotting interface

Plotting_ClassDiagram

### CODE REVIEW

File: src/gui/plugins/plotting/Plotting.cc
Comment (chapulina) [inline src/gui/plugins/plotting/Plotting.cc:28]: Missing `\brief`

File: src/gui/plugins/component_inspector/Vector3d.qml
Comment (chapulina) [inline src/gui/plugins/component_inspector/Vector3d.qml:90]: The `typeId` is convenient to uniquely identify each component type, but it isn't very user friendly on the UI. For example, it's hard for a user to understand that the pose x, y and z are being plotted here:

model_plot

Would it be possible to use `typeName` when displaying component information? The internal bookkeeping can remain using `typeId`.

File: src/gui/plugins/component_inspector/Vector3d.qml
Comment (chapulina) [inline src/gui/plugins/component_inspector/Vector3d.qml:82]: Be careful with indentation :wink:

File: src/gui/plugins/component_inspector/Pose3d.qml
Comment (chapulina) [inline src/gui/plugins/component_inspector/Pose3d.qml:192]: Can we keep the indentation? It helps communicate that the contents are iniside the header.

File: src/gui/plugins/plotting/Plotting.cc
Comment (chapulina) [inline src/gui/plugins/plotting/Plotting.cc:34]: I think you can use `Entity` instead of `uint64_t`. This way, if we ever change `Entity`, this will be updated at the same time. You don't need to change the signals and slots.

File: src/gui/plugins/plotting/Plotting.cc
Comment (chapulina) [inline src/gui/plugins/plotting/Plotting.cc:89]: You should check if `this->dataPtr->data.find(_attribute) != this->dataPtr->data.end()` before calling `AddChart`.

This also applies to other places.

File: src/gui/plugins/plotting/Plotting.hh
Comment (chapulina) [inline src/gui/plugins/plotting/Plotting.hh:31]: These includes can all be in the `.cc` file. It's preferable to include on the cc file whenever possible, because they won't get added to everyone who includes this header.

File: src/gui/plugins/plotting/Plotting.hh
Comment (chapulina) [inline src/gui/plugins/plotting/Plotting.hh:68]: Can you add a `brief` to this class explaining how it's supposed to be used?

File: src/gui/plugins/plotting/Plotting.hh
Comment (chapulina) [inline src/gui/plugins/plotting/Plotting.hh:124]:
```suggestion
  // Documentation inherited
```

No need to document the inherited functions unless you're adding special functionality to them.

File: src/gui/plugins/plotting/Plotting.hh
Comment (chapulina) [inline src/gui/plugins/plotting/Plotting.hh:162]: I think it would be helpful if your example also mentioned the string. For the "x attribute in Pose3d Component", what does `_attribute` look like?

File: src/gui/plugins/plotting/Plotting.qml
Comment (chapulina) [inline src/gui/plugins/plotting/Plotting.qml:1]: Missing copyright header.

Also, it would be nice to add minimum layout sizes.

File: src/gui/plugins/plotting/Plotting.cc
Comment (chapulina) [inline src/gui/plugins/plotting/Plotting.cc:209]:
```suggestion
  igndbg << "Unregister chart [" << Id << "]" << std::endl;
```

File: src/gui/plugins/plotting/Plotting.cc
Comment (chapulina) [inline src/gui/plugins/plotting/Plotting.cc:208]: Style

```suggestion
  std::string id = std::to_string(_entity) + "," + std::to_string(_typeId);
```

File: src/gui/plugins/plotting/Plotting.cc
Comment (AmrElsersy) [reply src/gui/plugins/plotting/Plotting.cc:89]: yeah i already do that, in that if condition above it (and all other places)

File: src/gui/plugins/component_inspector/Vector3d.qml
Comment (AmrElsersy) [reply src/gui/plugins/component_inspector/Vector3d.qml]: ok i made the modification for that PR
but the main code of that will be in the PlottingInterface code so i will update the other PR

**Review by claireyywang — COMMENTED** [review]
Features worked really well. Just a small comment. After it's resolved then I think this PR will be ready.
```
 $ ign gazebo levels.sdf
free(): invalid pointer
```
`invalid pointer` warning upon shutdown. I believe this is introduced by the `plotting.cc/hh` code after cross comparing with the master branch and testing the  `ComponentInspector` code only.

**Comment by AmrElsersy** [comment]
> Features worked really well. Just a small comment. After it's resolved then I think this PR will be ready.
> 
> ```
>  $ ign gazebo levels.sdf
> free(): invalid pointer
> ```
> 
> `invalid pointer` warning upon shutdown. I believe this is introduced by the `plotting.cc/hh` code after cross comparing with the master branch and testing the `ComponentInspector` code only.

File: src/gui/plugins/plotting/Plotting.cc
Comment (chapulina) [inline src/gui/plugins/plotting/Plotting.cc:62]: Do the `PlotComponent`s need to be pointers or can they just be objects? Then we don't need to worry about deleting them. I think that this memory isn't being cleared right now.

File: src/gui/plugins/plotting/Plotting.cc
Comment (chapulina) [inline src/gui/plugins/plotting/Plotting.cc:78]: The same thing here about using plan objects instead of pointers.

**Review by chapulina — COMMENTED** [review]
It's working for me, just have comments on some implementation details.

File: src/gui/plugins/plotting/Plotting.cc
Comment (AmrElsersy) [reply src/gui/plugins/plotting/Plotting.cc:48]: Yeah, i add them at run time and i PlotData class dosn't have Copy Constructor to assign local variables to it without dynamically allocating (will require another PR to ign-gui code)

anyway, i handled deleting them

**Review by chapulina — CHANGES_REQUESTED** [review]
There are some code checker issues :wink: Be sure to run `tools/code_check.sh`

```
  ./src/gui/plugins/plotting/Plotting.cc:20:  Found C++ system header after other header. Should be: Plotting.h, c system, c++ system, other.  [build/include_order] [4]
  ./src/gui/plugins/plotting/Plotting.cc:395:  Redundant blank line at the end of a code block should be deleted.  [whitespace/blank_line] [3]
  ./src/gui/plugins/plotting/Plotting.hh:126:  Lines should be <= 80 characters long  [whitespace/line_length] [2]
  ./src/gui/plugins/plotting/Plotting.hh:153:  Add #include <memory> for unique_ptr<>  [build/include_what_you_use] [4]
```

**Review by chapulina — APPROVED** [review]
Working for me! I pushed some minor changes on 095b1fedec0dd8bc0c89769ae85b596a3f02cca1 related to smart pointers and adding the plugin name.

Also opened https://github.com/ignitionrobotics/ign-gui/pull/125 with a few more tweaks.

Good to merge once CI is happy and @claireyywang has also approved.

**Review by claireyywang — APPROVED** [review]
LGTM!

**Pull request merged**
